crates/stages/stages/src/stages/headers.rs

```diff
@@ -155,11 +155,13 @@ where
 
         // If we only have the genesis block hash, then we are at first sync, and we can remove it,
         // add it to the collector and use tx.append on all hashes.
-        if let Some((hash, block_number)) = cursor_header_numbers.last()? {
-            if block_number.value()? == 0 {
-                self.hash_collector.insert(hash.key()?, 0)?;
-                cursor_header_numbers.delete_current()?;
-                first_sync = true;
+        if provider.tx_ref().entries::<RawTable<tables::HeaderNumbers>>()? == 1 {
```

Comment:
I'm not sure I understand how we can end up in a state where the original check would not be sufficient, can you expand on it a bit?

Reply:
yep, in my case we had a chain of hashes like this (just prefixes): [0xe, 0x2, 0xb, 0x9]

because `.last()` is returning the lexicographically highest hash in database, for all of those imports we eneded up with `first_sync = true` because we've always got the genesis hash (0xe)

this was fine for first 2 blocks because their hashes were increasing and the table keys evolved like [0xe] -> [] -> [0x2, 0xe] -> [0x2] -> [0x2, 0xb, 0xe]

however on the 0x9 block we tried to do [0x2, 0xb, 0xe] -> [0x2, 0xb] -> [0x2, 0xb, 0x9] which is incorrect because 0x9 should go before 0xb

hopefully this makes sense..

Reply:
ahh right i forgot this table is keyed by hash